lib/ecto/adapters/sojourn_broker/worker.ex

```diff
@@ -0,0 +1,230 @@
+defmodule Ecto.Adapters.SojournBroker.Worker do
+  @moduledoc false
+
+  use Connection
+  require Logger
+  use Bitwise
+  alias Ecto.Adapters.Pool
+
+  @timeout 5_000
+
+  @spec start_link(module, Keyword.t) :: {:ok, pid}
+  def start_link(module, params) do
+    Connection.start_link(__MODULE__, {module, params}, [])
+  end
+
+  @spec break(pid, reference, timeout) :: :ok
+  def break(worker, ref, timeout) do
+    Connection.call(worker, {:break, ref}, timeout)
+  end
+
+  @spec transaction_mode(pid, reference, Pool.mode, timeout) ::
+    :ok | {:error, :noconnect}
+  def transaction_mode(worker, ref, mode, timeout) do
+    Connection.call(worker, {:transaction_mode, ref, mode}, timeout)
+  end
+
+  @spec done(pid, reference) :: :ok
+  def done(worker, ref) do
+    Connection.cast(worker, {:done, ref})
+  end
+
+  ## Callbacks
+
+  def init({module, opts}) do
+    Process.flag(:trap_exit, true)
+    worker_keys = [:name, :min_backoff, :max_backoff]
+    {worker_opts, params} = Keyword.split(opts, worker_keys)
+    broker = Keyword.fetch!(worker_opts, :name)
+    min_backoff = Keyword.get(worker_opts, :min_backoff, 500)
+    max_backoff = Keyword.get(worker_opts, :max_backoff, 5_000)
+    backoff_threshold = div(max_backoff, 3)
+
+    s = %{conn: nil, module: module, params: params, transaction: nil,
+          broker: broker, tag: make_ref(), ref: nil, fun: nil, monitor: nil,
+          mode: :raw, backoff: min_backoff, min_backoff: min_backoff,
+          max_backoff: max_backoff, backoff_threshold: backoff_threshold}
+    {:connect, :init, s}
+  end
+
+  def connect(_, %{conn: nil, module: module, params: params,
+  min_backoff: min_backoff} = s) do
```

Comment:
You can have everything on the same line up to 80 chars. If it is too big, you can do the pattern matching inside the fun itself. :D